src/vm/method.cpp

```diff
@@ -4798,6 +4811,214 @@ Precode* MethodDesc::GetOrCreatePrecode()
+    // to ensure native slots are available where needed.
+    if (g_pConfig->TieredCompilation() &&
+        !IsZapped() &&
+        HasNativeCodeSlot() &&
+        !IsWrapperStub() &&
+        !IsDynamicMethod() &&
+        !GetLoaderAllocator()->IsCollectible() &&
+        !IsEnCMethod() &&
+        !CORDisableJITOptimizations(GetModule()->GetDebuggerInfoBits()) &&
+        !CORProfilerDisableTieredCompilation())
+    {
+        m_bFlags2 |= enum_flag2_IsEligibleForTieredCompilation;
+        _ASSERTE(IsTieredMethodVersionableWithPrecode() || IsTieredMethodVersionableWithVtableSlotBackpatch());
+        return true;
+    }
+#endif
+
+    return false;
+}
+
+#ifndef CROSSGEN_COMPILE
+
+void MethodDesc::RecordAndBackpatchEntryPointSlot(
+    LoaderAllocator *slotLoaderAllocator,
+    TADDR slot,
+    EntryPointSlotsToBackpatch::SlotType slotType)
+{
+    WRAPPER_NO_CONTRACT;
+
+    LoaderAllocator *mdLoaderAllocator = GetLoaderAllocator();
+    MethodDescBackpatchInfoTracker::ConditionalLockHolder lockHolder;
+
+    RecordAndBackpatchEntryPointSlot_Locked(
+        mdLoaderAllocator,
+        slotLoaderAllocator,
+        slot,
+        slotType,
+        GetEntryPointToBackpatch_Locked());
+}
+
+// This function tries to record a slot that would contain an entry point for the method, and backpatches the slot to contain
+// method's current entry point. Once recorded, changes to the entry point due to tiering will cause the slot to be backpatched
+// as necessary.
+void MethodDesc::RecordAndBackpatchEntryPointSlot_Locked(
+    LoaderAllocator *mdLoaderAllocator,
+    LoaderAllocator *slotLoaderAllocator,
+    TADDR slot,
+    EntryPointSlotsToBackpatch::SlotType slotType,
+    PCODE currentEntryPoint)
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(MethodDescBackpatchInfoTracker::IsLockedByCurrentThread());
+    _ASSERTE(mdLoaderAllocator != nullptr);
+    _ASSERTE(mdLoaderAllocator == GetLoaderAllocator());
+    _ASSERTE(slotLoaderAllocator != nullptr);
+    _ASSERTE(slot != NULL);
+    _ASSERTE(slotType < EntryPointSlotsToBackpatch::SlotType_Count);
+    _ASSERTE(IsTieredVtableMethod());
+
+    // The speciifed current entry point must actually be *current* in the sense that it must have been retrieved inside the
+    // lock, such that a recorded slot is guaranteed to point to the entry point at the time at which it was recorded, in order
+    // to synchronize with backpatching in MethodDesc::BackpatchEntryPointSlots(). If a slot pointing to an older entry point
+    // were to be recorded due to concurrency issues, it would not get backpatched to point to the more recent, actually
+    // current, entry point until another entry point change, which may never happen.
+    _ASSERTE(currentEntryPoint == GetEntryPointToBackpatch_Locked());
+
+    MethodDescBackpatchInfo *backpatchInfo =
+        mdLoaderAllocator->GetMethodDescBackpatchInfoTracker()->GetOrAddBackpatchInfo_Locked(this);
+    if (slotLoaderAllocator == mdLoaderAllocator)
```

Comment:
Currently for recorded slots, the slot's LoaderAllocator is either the same as the MethodDesc's, or it's a LoaderAllocator dependent on the MethodDesc, so the slot's LoaderAllocator's lifetime will be <= the MethodDesc's LoaderAllocator's lifetime. The MethodDesc can't disappear while there are still slots referring to its code. I'm not sure when it would be a common case.

Reply:
If A and B have AppDomain lifetime, why would they not use the same LoaderAllocator? Maybe they don't currently but perhaps that could be a better solution?

Reply:
I might be mistaken and maybe they already do, but I thought it was one per module.